### PR TITLE
Docs. Closes #4

### LINKED ISSUE #4: Add proper godoc
Once #1 is solved, and it is validated that no further API changes are required

### PR DESCRIPTION
**Pull request by ribasushi**
This is a lot of text (and code). This is essentially my first-ever standalone golang-based contribution.

Roast me! Anything goes: documentation style nits, naming nits, code nits. The library does seem to work flawlessly ( 48h very high cncurrency spin so far with no deadlocks/overwrites ).

/cc @warpfork as well ( can't add you as reviewer for some reason )

### CODE REVIEW

File: qringbuf.go
Comment (warpfork) [inline qringbuf.go:11]: s/Af/At/

File: qringbuf.go
Comment (dirkmc) [inline qringbuf.go:13]: Nit: It wasn't clear to me on the first reading that NextRegion() takes a parameter

File: qringbuf.go
Comment (dirkmc) [inline qringbuf.go:15]: I'm not sure what a "trailing" region means here. It might make sense to give a small example of the methods on the API and how they're typically used before the guarantees section, to give the reader context

File: qringbuf.go
Comment (dirkmc) [inline qringbuf.go:18]: nit: does intact mean "not get overwritten"?

File: qringbuf.go
Comment (dirkmc) [inline qringbuf.go:28]: Any reason to use int64 here instead of uint64?

File: qringbuf.go
Comment (dirkmc) [inline qringbuf.go:34]: Usually in go I'd expect to check `if streamErr != nil`, is there a reason to instead check `if reg == nil`?

File: qringbuf.go
Comment (dirkmc) [inline qringbuf.go:33]: I'd have to think a bit about why the parameter to NextRegion is `available - processed`. Could you add a comment explaining this?

File: qringbuf.go
Comment (dirkmc) [inline qringbuf.go:102]: Cool diagrams 👍 

I think this section may be a little clearer if it had an example showing the various function calls, and then the diagrams refer to those line numbers.

**Review by dirkmc — COMMENTED** [review]
Nice diagrams :)

I just had a look at the documentation sections to begin with, mostly nits there, I'll take a look at the code sections in a little bit

File: qringbuf.go
Comment (ribasushi) [reply qringbuf.go]: Addressed as https://github.com/ipfs/go-qringbuf/pull/6/commits/297eacd4921ad6
From commit:
> Did not use actual variables - when rendered I find it too distracting
> Happy to redo if a better suggestion comes along

File: qringbuf.go
Comment (ribasushi) [reply qringbuf.go:28]: I modeled this on the entirety of io.*, which never uses uint ( actually stdlib seems to avoid stdlib as a whole): https://golang.org/pkg/io
I assume this is to make it easier to detect underflows - just a bit too much subtracting and you are into "impossible negative number" territory, as opposed to 2^64-1

File: qringbuf.go
Comment (ribasushi) [reply qringbuf.go:102]: Thanks! Took way longer than I initially thought it will, but actually uncovered an inconsistency while I was drawing them, so all is well.

I am not entirely sure what you are suggesting. Perhaps discussing this voice some time today would help?

File: qringbuf.go
Comment (ribasushi) [reply qringbuf.go:36]: Attempted to address in https://github.com/ipfs/go-qringbuf/pull/6/commits/fbfe0db2fda82 Let me know if not sufficiently clear...

File: README.md
Comment (ribasushi) [reply README.md:14]: I was trying to put in one sentence the entire description of StartFill. Here is a punt of "go read there": https://github.com/ipfs/go-qringbuf/pull/6/commits/67a3224d1a78c
I can't think of a good way to distill this better. Suggestions welcome!

File: qringbuf.go
Comment (ribasushi) [reply qringbuf.go]: See https://github.com/ipfs/go-qringbuf/pull/6/commits/4d65af211de

File: qringbuf.go
Comment (ribasushi) [reply qringbuf.go]: See https://github.com/ipfs/go-qringbuf/pull/6/commits/4d65af211de

File: qringbuf.go
Comment (ribasushi) [reply qringbuf.go]: TLDR: this did end up making me change code, but the can of worms is strong with that one: https://github.com/ipfs/go-qringbuf/pull/6/commits/4d65af211de

Now the long version: This is how I wrote go at first too, and then got legit burned in some unrelated code, realizing **checking for error first/only is incorrect**. From https://golang.org/pkg/io/#Reader :

> When Read encounters an error or end-of-file condition after successfully reading n > 0 bytes, it returns the number of bytes read. ... Callers **should always process the n > 0 bytes returned before considering the error err**. Doing so correctly handles I/O errors that happen after reading some bytes and also both of the allowed EOF behaviors. 

In the example you are asking about, consider that `io.EOF` will start showing up the moment the collector reaches it ( and shuts down ). One will still cycle over the remaining, probably very large region, potentially multiple times requesting less and less of the remaining data: each of these calls will indicate `by the way, nothing extra is coming anymore`, but you are not in an actual "stop work" error condition.

On the other hand I did it wrong anyway: as it was, if the collector "eats" some data, and then the underlying stream errors: the data is now locked in the buffer without a way to get it out. So reworked as https://github.com/ipfs/go-qringbuf/pull/6/commits/4d65af211de

Comment (dirkmc) [reply qringbuf.go]: Ok makes sense - I think the important thing is to make clear any deviation from idiomatic patterns and the example comments & explanation you added address that 👍

File: qringbuf.go
Comment (dirkmc) [reply qringbuf.go:115]: What I mean is that the numbered items seem to be describing a piece of code:
⓪ Buffer initialized, StartFill(…) is called ...
① NextRegion(0) is blocked until ...

So it might be clearer if there's a concrete example of that code to compare with:
Line 3: Buffer initialized, StartFill(…) is called ...
Line 5: NextRegion(0) is blocked until ...

Happy to talk on the phone if that's not clear :)

Comment (ribasushi) [reply qringbuf.go:125]: Ohhhhh I see. This... wouldn't be... something that can be shown in docs, as the diagrams show the "tricky bits" which generally arise quite seldomly if ever. In fact the only way I managed to test all this is by essentially fuzzing the library by spinning it on an overloaded machine starved of CPU time.

I added a bit more prose in https://github.com/ipfs/go-qringbuf/commit/cb4888ddf8db, below is what it looks like visually. If I misunderstood what you are suggesting: let's chat after the launch meet
image

Comment (dirkmc) [reply qringbuf.go:125]: Ok no worries, I don't think you need the disclaimer - it was just a suggestion to add some context to the diagrams, but I think it's fine as is 👍

**Review by dirkmc — APPROVED** [review]

File: qringbuf.go
Comment (ribasushi) [reply qringbuf.go:125]: No, the disclaimer is good to have - if anything it puts things in perspective. So I'll def keep it
Thanks for all your suggestions, things are *WAY* better now ;)

**Comment by ribasushi** [comment]
I will keep this open for a bit longer if anyone else wants to chime in. Will merge Wed the 6th if I don't hear any other API suggestions.

**Comment by dirkmc** [comment]
My review was just for this PR - but it occurs to me now that you may have also wanted a review of all the code, is that right?

**Comment by ribasushi** [comment]
@dirkmc it certainly won't hurt, yes ;) But also if time is short - don't sweat it!

**Comment by ribasushi** [comment]
@dirkmc mainly I am interested in an in-depth "surface API" review. That is: when the code does something dumb, we can fix it. API - more difficult.

File: qringbuf.go
Comment (Stebalien) [inline qringbuf.go:94]: Avoid exposing mutexes like this. It makes internal locking patterns a part of the API.

File: qringbuf.go
Comment (Stebalien) [inline qringbuf.go:505]: I expect seeking backwards will be common, right? In that case, how about re-defining `NextRegion` as `NextRegion(offset)` where `offset` seeks _forward_ from the start of the previous region. Calling this with `0` would just return the previously read region. That should lead to fewer arithmetic errors.

File: qringbuf.go
Comment (Stebalien) [inline qringbuf.go:236]: If you want to be _really_ fast and sneaky, you could say that this is undefined behavior. If you did that, you'd be able to stick used regions into a `sync.Pool`.

**Review by Stebalien — COMMENTED** [review]
Nice! I actually like this API overall.

File: qringbuf.go
Comment (ribasushi) [reply qringbuf.go:827]: See https://github.com/ipfs/go-qringbuf/blob/e7dae312664e482074561c7c913605c6cdc58230/qringbuf.go#L833-L850
Essentially qringbuf does not care what your offset is within the reader. It just... goes forward.

This is also why I explicitly expose the mutex as stable part of the API - it allows me to "suspend everything" and jump way forward if I need to... though it now occurs to me that I need to remove the locking from Buffered() in order to make this workable.

Back to your questions above:
- Yes, otherwise NextRegion will say "dunno what to give you: a collector is not running"
- If you say 0, the only thing that changes is when the collector exits ( io.EOF ). Notably:
https://github.com/ipfs/go-qringbuf/blob/e7dae312664e482074561c7c913605c6cdc58230/qringbuf.go#L890-L894 in combination with https://github.com/ipfs/go-qringbuf/blob/e7dae312664e482074561c7c913605c6cdc58230/qringbuf.go#L769-L779

Suggestion how to make this... clearer?

Comment (Stebalien) [reply qringbuf.go:827]: > This is also why I explicitly expose the mutex as stable part of the API - it allows me to "suspend everything" and jump way forward if I need to... though it now occurs to me that I need to remove the locking from Buffered() in order to make this workable.

Why do you need to _suspend_ it? That seems very strange. If you need a way to "pause", just expose a separate method and automatically resume when the user next asks for a region.

> Yes, otherwise NextRegion will say "dunno what to give you: a collector is not running"

Why not just automatically start when the user asks for a region? Personally, I'd expect the API to be:

```go
qrb.SetLimit(N) // set the limit on the number of bytes to read relative to the current offset.
```

---

Basically, forget about the internals and try to figure out what you'd like the API to look like regardless of how this library is implemented under the hood.

Why would users care about explicitly starting/stopping the collector? Why does the user even care that there is a "collector".

The user cares about:

* Closing the buffer when done with it.
* Reading slices from the buffer.
* Setting limits (not actually sure _why_?).

File: qringbuf.go
Comment (ribasushi) [reply qringbuf.go]: I did and added the spaces afterwards, because of it rendering weird in a browser. The fact that you noticed this, means it is jarring for others too, so reverting all the post-formatting 😿 

https://github.com/ipfs/go-qringbuf/pull/6/commits/ee59439441fd

File: qringbuf.go
Comment (ribasushi) [reply qringbuf.go]: I don't see how... that can work? We need to chat about this more...

An aside: I did have an implementation with sync.Pool in dagger itself over millions of 9-byte slices, where the freeing is definitive and thus a Pool makes sense -  I did rip it all out, because while allocations went down in a profile, there was **no actual difference, not even sub-%, outright undetectable**

Spoke with Eric a bit, it seems the compiler just got far enough along that sync.Pool is becoming needless overhead.

File: qringbuf.go
Comment (ribasushi) [reply qringbuf.go]: We need to chat about this more...

At present the panic of "you gave me back everything I gave you" has saved me from countless endless-loops. Having to keep track of "what did I eat" and advancing your own "view" by hand seems... hrm.

As a general rule seeking back a super-short amount is common ( due to your last record having been cut ). Re-reading the entire buffer though is outright a bug.

File: qringbuf.go
Comment (ribasushi) [reply qringbuf.go]: I guess I need to somehow make it clearer that a reader is not exclusive to the qringbuf, it merely cooperates. Will do so in the final commit to push.

I guess... what about changing doc above to say "if you want me to stop - retain and close the underlying reader", and removing the StopFill() method entirely ( thus reducing complexity way down ).

Actually **I love that**... here is how much crap this throws away: https://github.com/ipfs/go-qringbuf/commit/8eaa25fd2947

One extra doc update coming in a separate commit

Comment (ribasushi) [reply qringbuf.go]: And here is the doc part: https://github.com/ipfs/go-qringbuf/blob/9f05222fd67da74107b91461c9f1d3c97e160ce3/qringbuf.go#L814-L817

File: qringbuf.go
Comment (ribasushi) [reply qringbuf.go]: THANK YOU! This is exactly the type of review/comment-thread I was looking for. Sorry for rewriting half the docs - one would need to re-read this almost from scratch.

https://github.com/ipfs/go-qringbuf/commit/9f05222fd67 resolves all remaining comments ( I marked them as such ). Notably:

> Can we put this in the description for the individual methods? The reader won't know what any of these functions do till they read the actual API.

Re-wrapped, I did a bit of both. Worry is that the reader **won't read** down to the API without an overview making them go "huh... that's... not something I've seen before..."

> I'd focus more on why and what and less on how.

Tried to do so... not sure how successful. Let's chat about this more ;)

> Avoid exposing mutexes 

This is explicitly by design - expanded upon here: https://github.com/ipfs/go-qringbuf/blob/9f05222fd67da74107b91461c9f1d3c97e160ce3/qringbuf.go#L8-L11 and here https://github.com/ipfs/go-qringbuf/blob/9f05222fd67da74107b91461c9f1d3c97e160ce3/qringbuf.go#L519-L521

File: qringbuf.go
Comment (abo-L) [inline qringbuf.go:13]: TON

**Review by abo-L — CHANGES_REQUESTED** [review]

**Review by abo-L — APPROVED** [review]

**Pull request merged**
